src/nu/validator/htmlparser/impl/TreeBuilder.java

```diff
@@ -1458,6 +1460,17 @@ public final void startTag(ElementName elementName,
         flushCharacters();
 
         // [NOCPP[
+        if (contextNamespace == "http://www.w3.org/1999/xhtml" &&
+                selfClosing && errorProfileMap != null &&
+                errorProfileMap.get("html-strict") != null) {
+            warn("Self-closing tag syntax in text/html documents is widely"
+                    + " discouraged; it’s unnecessary and interacts badly"
+                    + " with other HTML features (e.g., unquoted attribute"
+                    + " values). If you’re using a tool that injects"
+                    + " self-closing tag syntax into all void elements,"
+                    + " without any option to prevent it from doing so,"
+                    + " then consider switching to a different tool.");
+        }
```

Comment:
This now emits a duplicate message when there is an actual error here:
https://searchfox.org/mozilla-central/rev/929b2a7154e463674ebac497d4a89208eec1b8f7/parser/html/javasrc/TreeBuilder.java#2846

To get this right, I think you need two booleans initialized at to top of the method in a `NOCPP` block: `boolean wasSelfClosing = selfClosing;` and `boolean voidElement = false;` Then the non-foreign cases that set `selfClosing = false;` should, in a `NOCPP` block also set `voidElement = true;`. This should then be enough to decide if a non-error is appropriate as an `else` branch to the check for `selfClosing` at the end of the method.

Reply:
> To get this right, I think you need two booleans initialized at to top of the method in a `NOCPP` block: `boolean wasSelfClosing = selfClosing;` and `boolean voidElement = false;` Then the non-foreign cases that set `selfClosing = false;` should, in a `NOCPP` block also set `voidElement = true;`. This should then be enough to decide if a non-error is appropriate as an `else` branch to the check for `selfClosing` at the end of the method.

Thanks, did it that way